File: integration/src/test/java/my/sample/integration/SampleSpec.java
package my.sample.integration;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.ops4j.pax.exam.junit.ExamReactorStrategy;
import org.ops4j.pax.exam.junit.JUnit4TestRunner;
import org.ops4j.pax.exam.spi.reactors.AllConfinedStagedReactorFactory;

import javax.management.MBeanServerConnection;
import javax.management.ObjectName;
import javax.management.openmbean.TabularData;
import javax.management.remote.JMXConnector;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

/**
 * @author mohammad shamsi <m.h.shams@gmail.com>
 */

@RunWith(JUnit4TestRunner.class)
@ExamReactorStrategy(AllConfinedStagedReactorFactory.class)
public class SampleSpec extends BaseKarafSupport {

    @Test
    public void listCommand() throws Exception {
        String listOutput = executeCommand("features:list");
        System.out.println(listOutput);
        assertFalse(listOutput.isEmpty());
        listOutput = executeCommand("features:list -i");
        System.out.println(listOutput);
        assertFalse(listOutput.isEmpty());
    }

    @Test
    public void listViaMBean() throws Exception {
        JMXConnector connector = null;
        try {
            connector = this.getJMXConnector();
            MBeanServerConnection connection = connector.getMBeanServerConnection();
            ObjectName name = new ObjectName("org.apache.karaf:type=features,name=root");
            TabularData features = (TabularData) connection.getAttribute(name, "Features");
            assertTrue(features.size() > 0);
        } finally {
            if (connector != null)
                connector.close();
        }
    }

    @Test
    public void repoAddRemoveCommand() throws Exception {
        String featureUrl = "mvn:my.sample/feature/1.1-SNAPSHOT/xml/features";

        //Add feature repository
        System.out.println(executeCommand("features:addurl " + featureUrl));
        String repoListOutput = executeCommand("features:listurl");
        System.out.println(repoListOutput);
        assertTrue(repoListOutput.contains(featureUrl));

        //Install feature
        String featureInstallOutput = executeCommand("features:install -v feature", 3 * 60 * 1000L, false);
        System.out.println(featureInstallOutput);
        assertFalse(featureInstallOutput.isEmpty());

        //Check if feature is installed
        String listOutput = executeCommand("features:list -i");
        System.out.println(listOutput);
        assertTrue(listOutput.contains("feature"));

        //Check bundles
        String bundlesOutput = executeCommand("osgi:list");
        System.out.println(bundlesOutput);
    }
}

File: integration/src/test/kotlin/my/sample/integration/SampleSpec.kt
package my.sample.integration

import org.junit.Test as test
import org.junit.runner.RunWith as runWith
import org.ops4j.pax.exam.junit.ExamReactorStrategy as examReactorStrategy
import org.ops4j.pax.exam.junit.JUnit4TestRunner
import org.ops4j.pax.exam.spi.reactors.AllConfinedStagedReactorFactory

import javax.management.ObjectName
import javax.management.openmbean.TabularData
import javax.management.remote.JMXConnector
import org.junit.Assert


/**
 * @author mohammad shamsi <m.h.shams@gmail.com>
 */

runWith(javaClass<JUnit4TestRunner>())
examReactorStrategy(javaClass<AllConfinedStagedReactorFactory>())
public class SampleSpec : BaseKarafSupport() {

    test
    public fun listCommand() {
        var listOutput = executeCommand("features:list")
        System.out.println(listOutput)
        Assert.assertFalse(listOutput.isEmpty())
        listOutput = executeCommand("features:list -i")
        println(listOutput)
        Assert.assertFalse(listOutput.isEmpty())
    }

    test
    public fun listViaMBean() {
        var connector: JMXConnector? = null
        try {
            connector = this.getJMXConnector()
            val connection = connector?.getMBeanServerConnection()
            val name = ObjectName("org.apache.karaf:type=features,name=root")
            val features = connection?.getAttribute(name, "Features") as TabularData
            Assert.assertTrue(features.size() > 0)
        } finally {
            if (connector != null)
                connector?.close()
        }
    }

    test
    public fun repoAddRemoveCommand() {
        val featureUrl = "mvn:my.sample/feature/1.1-SNAPSHOT/xml/features"

        //Add feature repository
        println(executeCommand("features:addurl $featureUrl"))
        val repoListOutput = executeCommand("features:listurl")
        println(repoListOutput)
        Assert.assertTrue(repoListOutput.contains(featureUrl))

        //Install feature
        val featureInstallOutput = executeCommand("features:install -v feature", 3 * 60 * 1000L, false)
        println(featureInstallOutput)
        Assert.assertFalse(featureInstallOutput.isEmpty())

        //Check if feature is installed
        val listOutput = executeCommand("features:list -i")
        println(listOutput)
        Assert.assertTrue(listOutput.contains("feature"))

        //Check bundles
        val bundlesOutput = executeCommand("osgi:list")
        println(bundlesOutput)
    }
}
